### PR TITLE
gh-99593: Add tests for Unicode C API (part 2)

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Add tests for lower-level functions.

* Issue: gh-99593

### CODE REVIEW

File: Lib/test/test_capi/test_unicode.py
Comment (vstinner) [inline Lib/test/test_capi/test_unicode.py]: PyUnicode_ClearFreeList was removed in Python 3.9!

File: Lib/test/test_capi/test_unicode.py
Comment (vstinner) [inline Lib/test/test_capi/test_unicode.py:29]: Why not testing with ``MAX_UNICODE = 0x10ffff``, check that MAX_UNICODE is ok, but MAX_UNICODE+1 raises an error?

File: Lib/test/test_capi/test_unicode.py
Comment (vstinner) [inline Lib/test/test_capi/test_unicode.py:30]: Maybe this error should become a ValueError, but it can be changed outside this PR, since you seem to want to backport these news tests.

File: Modules/_testcapi/unicode.c
Comment (vstinner) [inline Modules/_testcapi/unicode.c:148]: It sounds dangerous to return a string to the "Python space" if characters are not initialized when maxchar is greated than 0x10ffff. Can you remove ``maxchar <= 0x10ffff`` condition? PyUnicode_Fill() must fail if the *fill* character is too big (greater than 0x10ffff).

File: Modules/_testcapi/unicode.c
Comment (vstinner) [inline Modules/_testcapi/unicode.c:201]: For me, -1 means an error. You shouldn't have to check if an exception was raised or not. Or it should become: assert(PyErr_Occurred()).

Same remark for other test wrapper like the one for PyUnicode_Resize().

File: Modules/_testcapi/unicode.c
Comment (vstinner) [inline Modules/_testcapi/unicode.c]: You should check left_copy is not NULL, maybe with: ``assert(left_copy != NULL)``.

Same remark for PyUnicode_AppendAndDel() below.

File: Modules/_testcapi/unicode.c
Comment (vstinner) [inline Modules/_testcapi/unicode.c]: If left_copy is not NULL, it sounds strange to me to call Py_DECREF() and then ``return left_copy``. The function should set left_copy to NULL on error. Either use ``assert(left_copy == NULL)`` or ``Py_CLEAR(left_copy)``.

File: Modules/_testcapi/unicode.c
Comment (vstinner) [inline Modules/_testcapi/unicode.c]: It's surprising to pass ``&size`` twice. Maybe add a comment to explain this uncommon code, something like: "If the size argument is set, override the 's' bytes string length".

For me, it would be more explicit if this C wrapper would always have two arguments and the test itself sets the argument argument to len(string). But you can leave it as it is.

File: Modules/_testcapi/unicode.c
Comment (vstinner) [inline Modules/_testcapi/unicode.c]: Can't you leave them uninitialized? Compilers complain about uninitialized values? Stupid compilers :-) I mean, this test is not about testing PyArg_Parse() :-) Well, it's just a remark, your code is fine.

File: Modules/_testcapi/unicode.c
Comment (vstinner) [inline Modules/_testcapi/unicode.c:425]: You might want to raise an error if bsize is not a multiple of SIZEOF_WCHAR_T.

File: Modules/_testcapi/unicode.c
Comment (vstinner) [inline Modules/_testcapi/unicode.c]: PyMem_FREE is a deprecated alias to PyMem_Free(). I would expect the opposite change, replace PyMem_FREE with PyMem_Free :-)

File: Lib/test/test_capi/test_unicode.py
Comment (vstinner) [inline Lib/test/test_capi/test_unicode.py:107]: Please add a comment to note that all strings have exactly 3 characters.

File: Lib/test/test_capi/test_unicode.py
Comment (vstinner) [inline Lib/test/test_capi/test_unicode.py:149]: That's done automatically by Refleaks buildbots, no?

File: Lib/test/test_capi/test_unicode.py
Comment (vstinner) [inline Lib/test/test_capi/test_unicode.py]: This condition looks wrong. Maybe just use ``else:``? I don't think that Python supports other endianness. Same a few lines below.

File: Lib/test/test_capi/test_unicode.py
Comment (serhiy-storchaka) [reply Lib/test/test_capi/test_unicode.py]: Oh, these tests were written when it was here.

File: Lib/test/test_capi/test_unicode.py
Comment (serhiy-storchaka) [reply Lib/test/test_capi/test_unicode.py:29]: Added test for 0x10ffff.

File: Lib/test/test_capi/test_unicode.py
Comment (serhiy-storchaka) [reply Lib/test/test_capi/test_unicode.py:30]: I think that SystemError is better exception type here. It is a misuse of the C API, you cannot get this error from Python code.

File: Lib/test/test_capi/test_unicode.py
Comment (serhiy-storchaka) [reply Lib/test/test_capi/test_unicode.py:107]: Done.

File: Lib/test/test_capi/test_unicode.py
Comment (serhiy-storchaka) [reply Lib/test/test_capi/test_unicode.py:149]: Maybe. But since this C API does unusual things with reference counts, it would be better to test it explicitly, if possible.

File: Lib/test/test_capi/test_unicode.py
Comment (serhiy-storchaka) [reply Lib/test/test_capi/test_unicode.py]: Good catch. I though that explicit check can be better, but made an error in it.

File: Modules/_testcapi/unicode.c
Comment (serhiy-storchaka) [reply Modules/_testcapi/unicode.c:148]: It is a test for `PyUnicode_New()`, not for `PyUnicode_Fill()`. We should see exceptions raised by `PyUnicode_New()`, not `PyUnicode_Fill()`.

It never happens, because `PyUnicode_New()` returns NULL if  size > 0 and maxchar > 0x10ffff. If it will not return NULL, it is better to get a malformed string than get an exception raised by `PyUnicode_Fill()` and think that it was raised by `PyUnicode_New()`.

File: Modules/_testcapi/unicode.c
Comment (serhiy-storchaka) [reply Modules/_testcapi/unicode.c:201]: If an exception was raised, but result is not -1, then what? The code will return NULL, and we will never know that something wrong happened with `PyUnicode_WriteChar()`.

I do not use the C assert() in these tests, because I want to make tests working even in non-debug build.

File: Modules/_testcapi/unicode.c
Comment (serhiy-storchaka) [reply Modules/_testcapi/unicode.c]: Agreed, this code is redundant. Removed it.

File: Modules/_testcapi/unicode.c
Comment (serhiy-storchaka) [reply Modules/_testcapi/unicode.c]: Rewritten in more explicit way, as in `unicode_fromwidechar()`.

File: Modules/_testcapi/unicode.c
Comment (serhiy-storchaka) [reply Modules/_testcapi/unicode.c]: Removed. I think it was just copied from other place.

File: Modules/_testcapi/unicode.c
Comment (serhiy-storchaka) [reply Modules/_testcapi/unicode.c:425]: Done.

File: Modules/_testcapi/unicode.c
Comment (serhiy-storchaka) [reply Modules/_testcapi/unicode.c]: Reverted. The code in my branch was written a long time ago, when PyMem_FREE and PyMem_Free were different things. I just copied it over the current code.

**Comment by serhiy-storchaka** [comment]
Thank you for your review Victor.

File: Modules/_testcapi/unicode.c
Comment (vstinner) [reply Modules/_testcapi/unicode.c:201]: > I do not use the C assert() in these tests, because I want to make tests working even in non-debug build.

C assert() are enabled on release (no-debug) builds: see the following code in Modules/_testcapi/parts.h:

```
// Always enable assertions
#undef NDEBUG
```

Many _testcapi tests are only implemented with ``assert()``.

**Review by vstinner — APPROVED** [review]
LGTM. Thanks for my addressing my review.

**Comment by hugovk** [comment]
@serhiy-storchaka Victor approved this, ready to merge?

(I removed the 3.10 backport label, it's now security only)

**Review by vstinner — APPROVED** [review]
LGTM

**Comment by vstinner** [comment]
The CI test "DO-NOT-MERGE / unresolved review" was blocked on "Waiting for status to be reported". I tried to update the PR on main to see if it does unblock the PR.

**Comment by vstinner** [comment]
Merged, thanks @serhiy-storchaka.

@serhiy-storchaka: If you consider that this change should be backported to Python 3.11, go ahead. But the automated backport failed for an unknown reason.

**Pull request merged**
